File: backend/adapters/springboot-jakarta/src/main/java/com/compounding/adapter/springboot/contract/WireResponse.java
package com.compounding.adapter.springboot.contract;

import com.fasterxml.jackson.annotation.JsonInclude;
import org.springframework.http.ResponseEntity;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Builder for wire-protocol-compliant response bodies.
 * Success: payload fields at top level, no "error" key (or null).
 * Failure: { "error": { "code", "message", "details"? } }
 *
 * Keeps response construction in one place rather than scattered across controllers.
 */
public final class WireResponse {

    private WireResponse() {}

    /** Build a success ResponseEntity from a plain field map. */
    public static ResponseEntity<Map<String, Object>> ok(Map<String, Object> fields) {
        return ResponseEntity.ok(fields);
    }

    /**
     * Build an error ResponseEntity.
     * HTTP status is resolved from ContractLoader (driven by codes.yaml).
     */
    public static ResponseEntity<Map<String, Object>> error(
            ContractLoader loader, String code) {
        return error(loader, code, null);
    }

    public static ResponseEntity<Map<String, Object>> error(
            ContractLoader loader, String code, Map<String, Object> details) {

        int httpStatus = loader.httpStatusFor(code);
        String message = loader.messageFor(code);

        ErrorEnvelope envelope = details != null
            ? ErrorEnvelope.withDetails(code, message, details)
            : ErrorEnvelope.of(code, message);

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("error", envelope);

        return ResponseEntity.status(httpStatus).body(body);
    }
}

File: backend/adapters/fastapi/wire_response.py
"""
wire_response.py — Wire-protocol-compliant response builders.

Success: payload fields at top level, no 'error' key.
Failure: { "error": { "code", "message", "details"? } }

Mirrors WireResponse.java + ErrorEnvelope.java behavior.
HTTP status for errors is resolved via ContractLoader (from codes.yaml — G-1).
"""

from __future__ import annotations

from typing import Any

from fastapi import Response
from fastapi.responses import JSONResponse

from contract_loader import contract


def ok(fields: dict[str, Any], status_code: int = 200) -> JSONResponse:
    """Return a success JSONResponse. No 'error' key in the body."""
    return JSONResponse(content=fields, status_code=status_code)


def error(code: str, details: dict[str, Any] | None = None) -> JSONResponse:
    """
    Return an error JSONResponse.
    HTTP status resolved from codes.yaml via ContractLoader (G-1 compliant).
    """
    http_status = contract.http_status_for(code)
    message = contract.message_for(code)

    envelope: dict[str, Any] = {"code": code, "message": message}
    if details is not None:
        envelope["details"] = details

    return JSONResponse(content={"error": envelope}, status_code=http_status)
